x/streamer/keeper/pump_stream.go

```diff
@@ -0,0 +1,370 @@
+
+	randomBig := new(big.Int).SetBytes(seed)
+	return new(big.Int).Mod(randomBig, modulo)
+}
+
+// ExecutePump performs the pump operation by buying tokens for a specific rollapp.
+// CONTRACT: pumpAmt is always in base denom.
+func (k Keeper) ExecutePump(
+	ctx sdk.Context,
+	pumpAmt sdk.Coin,
+	rollappID string,
+) (tokenOut sdk.Coin, err error) {
+	rollapp, found := k.rollappKeeper.GetRollapp(ctx, rollappID)
+	if !found {
+		return sdk.Coin{}, fmt.Errorf("rollapp not found: %s", rollappID)
+	}
+
+	plan, found := k.iroKeeper.GetPlanByRollapp(ctx, rollapp.RollappId)
+	if !found {
+		return sdk.Coin{}, fmt.Errorf("IRO plan not found for rollapp: %s", rollappID)
+	}
+
+	buyer := k.ak.GetModuleAddress(types.ModuleName)
+
+	var tokenOutDenom string
+	if plan.IsSettled() {
+		tokenOutDenom = plan.SettledDenom
+	} else {
+		tokenOutDenom = plan.LiquidityDenom
+	}
+	var tokenOutAmt math.Int
+
+	err = osmoutils.ApplyFuncIfNoError(ctx, func(ctx sdk.Context) error {
+		// There are several cases:
+		// - targetDenom == plan.SettledDenom =>
+		//     IRO is settled =>
+		//     RA token exists =>
+		//     every RA token has an AMM route to DYM =>
+		//     do an AMM swap using feetoken
+		//
+		// - targetDenom == plan.LiquidityDenom != BaseDenom =>
+		//     IRO is not settled =>
+		//     need to buy IRO using liquidity tokens =>
+		//     every liquidity token has an AMM route to DYM =>
+		//     do an AMM swap using feetoken =>
+		//     buy IRO using tokenOut of liquidity tokens
+		//
+		// - targetDenom == plan.LiquidityDenom == BaseDenom =>
+		//     IRO is not settled =>
+		//     need to buy IRO using liquidity tokens =>
+		//     pump amount is already in base denom =>
+		//     buy IRO using pumpAmt of BaseDenom
+
+		if tokenOutDenom == pumpAmt.Denom {
+			// This can happen only for unsettled IRO with LiquidityToken == BaseDenom
+			tokenOutAmt = pumpAmt.Amount
+		} else {
+			feeToken, err := k.txFeesKeeper.GetFeeToken(ctx, tokenOutDenom)
+			if err != nil {
+				return fmt.Errorf("get fee token for denom %s: %w", tokenOutDenom, err)
+			}
+
+			reverseRoute := reverseInRoute(feeToken.Route, tokenOutDenom)
+			tokenOutAmt, err = k.poolManagerKeeper.RouteExactAmountIn(
+				ctx,
+				buyer,
+				reverseRoute,
+				pumpAmt,        // token in
+				math.ZeroInt(), // no slippage
+			)
+			if err != nil {
+				return fmt.Errorf("route exact amount in: target denom: %s, error: %w", tokenOutDenom, err)
+			}
+		}
+
+		if !plan.IsSettled() {
+			// If IRO is in progress, use liquidity tokens to buy IRO tokens
+			tokenOutAmt, err = k.iroKeeper.BuyExactSpend(
+				ctx,
+				fmt.Sprintf("%d", plan.Id),
+				buyer,
+				tokenOutAmt,    // amountToSpend
+				math.ZeroInt(), // no slippage
+			)
+			if err != nil {
+				return fmt.Errorf("buy from IRO %d: %w", plan.Id, err)
+			}
+			// In that case, tokenOut is an IRO token
+			tokenOutDenom = plan.GetIRODenom()
+		}
+
+		return nil
+	})
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
```

Comment:
are we not missing the case when the plan rollapp token is already in an AMM and buying from there the token?